### PR TITLE
systemd: don't taint on unmerged /usr

### LINKED ISSUE #184826: systemd warns about `unmerged-usr` on NixOS
`systemctl status` started showing `Tainted: unmerged-usr` on NixOS (https://github.com/systemd/systemd/pull/22984), because it thinks NixOS has an unmerged `/usr`, but this concept doesn't apply to NixOS (everything is merged under `/nix/store`, we only keep `/usr/bin/env` and `/bin/sh` for compatibility).

There seem to be plans for "removing support for unmerged-usr" in systemd, so this should probably be addressed, although I'm not sure to what extent this "support" affects NixOS.

Possible solutions:

1. change the NixOS activation script to put both `sh` and `env` in `/usr/bin` and then symlink `/bin` to `/usr/bin`. This is quick and relatively easy to do (migrating existing setups might be a problem), but introduces more "impure" paths than strictly necessary (`/bin/env` and `/usr/bin/sh`).
2. patch systemd in nixpkgs (or just in NixOS?) to remove that flag/check/concept.
3. patch systemd upstream to detect such situations (maybe simply based on os-release?) and not complain.

I'll create an issue at systemd as well.

Pinging people I saw talk about this: @K900 @alyssais @lheckemann

### PR DESCRIPTION
**Pull request by oxalica**
###### Description of changes

Fixes #184826
Discussion: https://github.com/systemd/systemd/issues/24191#issuecomment-1209350080

~I'm still running the build. It may take some time.~
`systemd` builds and works as expected.

r? @flokli

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/os-specific/linux/systemd/0018-core-don-t-taint-on-unmerged-usr.patch [hunk 1]
```diff
@@ -0,0 +1,30 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: oxalica <oxalicc@pm.me>
+Date: Tue, 4 Oct 2022 09:18:07 +0800
+Subject: [PATCH] core: don't taint on unmerged /usr
+
+NixOS has very different approach towards /bin and /sbin - they don't really exist (except for /bin/sh and /usr/bin/env, because these are used heavily in shebangs around the world).
+The concept of merged or unmerged usr doesn't really apply here at all, it's neither of the two.
+Users don't execute things from /bin or /sbin, there's nothing else in there.
+In all cases, systemd doesn't look things up from /usr/bin or /bin, so showing the taint isn't really helpful.
```
Comment (flokli) [inline pkgs/os-specific/linux/systemd/0018-core-don-t-taint-on-unmerged-usr.patch:9]: Can you reflow that patch to a less large line width?

**Review by flokli — COMMENTED** [review]
Overall LGTM. Can you add a test to not see `unmerged-usr` in a "Tainted:" line in `systemctl status` output to `nixosTests.systemd`?

File: pkgs/os-specific/linux/systemd/0018-core-don-t-taint-on-unmerged-usr.patch (hunk 1 above)
Comment (oxalica) [reply pkgs/os-specific/linux/systemd/0018-core-don-t-taint-on-unmerged-usr.patch]: Fixed.

**Comment by flokli** [comment]
Thanks!

**Pull request merged**